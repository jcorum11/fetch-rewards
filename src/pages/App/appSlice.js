import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import appApi from '../../api/app'

export const sendInputs = createAsyncThunk('app/sendInputsStatus', async (inputs) => {
  const response = await appApi.sendInputs(inputs)
  return response
})

const appSlice = createSlice({
  name: 'app',
  initialState: {
    response: '',
    status: 'Idle...'
  },
  reducers: {},
  extraReducers: (builder) => {
    builder.addCase(sendInputs.pending, (state, action) => {
      state.status = 'Loading...'
    })
    builder.addCase(sendInputs.fulfilled, (state, action) => {
      state.status = 'Sent!'
      state.response = action.payload
    })
  }
})

export const selectResponse = state => state.app.response
export const selectPostStatus = state => state.app.status

export default appSlice.reducer
