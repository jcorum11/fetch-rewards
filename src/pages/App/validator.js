const validate = (payload) => {
  let result
  switch (true) {
    case !payload.name.first:
      result = false
      break
    case !payload.name.last:
      result = false
      break
    case !payload.name.middle:
      result = false
      break
    case !payload.password:
      result = false
      break
    case !payload.email:
      result = false
      break
    case !payload.occupation:
      result = false
      break
    case !payload.state:
      result = false
      break
    default:
      result = true
      break
  }
  return result
}

export default validate
