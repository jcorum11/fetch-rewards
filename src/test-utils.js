import React from 'react'
import { render as rtlRender } from '@testing-library/react'
import { configureStore } from '@reduxjs/toolkit'
import { Provider } from 'react-redux'
// Import your own reducer
import dropdownReducer from './components/DropdownInput/dropdownInputSlice'
import textInputReducer from './components/TextInput/textInputSlice'
import appReducer from './pages/App/appSlice'

function render(
  ui,
  {
    preloadedState,
    store = configureStore({ reducer: { dropdown: dropdownReducer, textInput: textInputReducer, app: appReducer }, preloadedState }),
    ...renderOptions
  } = {}
) {
  function Wrapper({ children }) {
    return <Provider store={store}>{children}</Provider>
  }
  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions })
}

// re-export everything
export * from '@testing-library/react'
// override render method
export { render }