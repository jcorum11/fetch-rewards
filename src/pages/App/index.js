/** @jsxImportSource @emotion/react */
import DropdownInput from '../../components/DropdownInput'
import TextInput from '../../components/TextInput'
import { useDispatch, useSelector } from 'react-redux'
import { useEffect, useState } from 'react'
import { fetchOptions, selectOccupation, selectOptionsFetchStatus, selectState } from '../../components/DropdownInput/dropdownInputSlice'
import { selectEmail, selectName, selectPassword } from '../../components/TextInput/textInputSlice'
import { selectPostStatus, sendInputs } from './appSlice'
import validate from './validator'
import { css } from '@emotion/react'

function App() {
  const dispatch = useDispatch()
  const optionStatus = useSelector(selectOptionsFetchStatus)
  const name = useSelector(selectName)
  const password = useSelector(selectPassword)
  const email = useSelector(selectEmail)
  const occupation = useSelector(selectOccupation)
  const state = useSelector(selectState)
  const postStatus = useSelector(selectPostStatus)
  useEffect(() => {
    dispatch(fetchOptions())
  }, [])
  const [validationFailed, setValidationFailed] = useState(false)
  const handleSubmit = () => {
    const payload = { name, password, email, occupation, state }
    console.log(payload)
    let result = validate(payload)
    console.log(result)
    if (validate(payload) === false) {
      setValidationFailed(true)
      return
    }
    setValidationFailed(false)
    dispatch(sendInputs(payload))
  }
  return (
    <div className="App" css={app}>
      <div css={container}>
        <header css={header}>
          <h1>Fetch Rewards Project</h1>
        </header>
        <TextInput type='Name' nameType='First' />
        <TextInput type='Name' nameType='Middle' />
        <TextInput type='Name' nameType='Last' />
        <TextInput type='Email' />
        <TextInput type='Password' />
        {optionStatus === 'complete' && <DropdownInput type='State' />}
        {optionStatus === 'complete' && <DropdownInput type='Occupation' />}
        <div css={buttonContainer}><button css={button} onClick={handleSubmit}>Submit</button></div>
        {validationFailed && <div css={messageContainer}><p css={message}>You didn't complete some fields!</p></div>}
        <div>
          <h1 css={header}>{postStatus}</h1>
        </div>
      </div>
    </div>
  )
}

const header = css`
display: flex;
justify-content: center;
color: white;
`

const container = css`
margin: 10vh 20vw 0;
background-color: darkcyan;
border-radius: 1rem;
width: 15vw;
`

const app = css`
display: flex;
justify-content: center;
`

const button = css`
color: darkcyan;
background-color: white;
height: 3rem;
width: 6rem;
border-radius: 10px;
font-weight: bold;
font-size: 15px;
`

const buttonContainer = css`
width: 100%;
display: flex;
justify-content: center;
margin: 1rem 0;
`

const messageContainer = css`
width: 100%;
display: flex;
justify-content: center;
`

const message = css`
color: white;
`

export default App
