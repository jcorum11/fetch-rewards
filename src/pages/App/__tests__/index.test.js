import React from 'react'
import '@testing-library/jest-dom/extend-expect'
import { render, cleanup } from '../../../test-utils'
import App from '..'

afterEach(cleanup)

describe('App', () => {
  it('Renders', () => {
    render(<App />)
  })

  it('Matches snapshot', () => {
    const { asFragment } = render(<App />)
    expect(asFragment()).toMatchSnapshot()
  })
})