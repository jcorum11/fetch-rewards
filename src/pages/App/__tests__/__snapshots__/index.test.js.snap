// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="App css-3vzvwk-App"
  >
    <div
      class="css-1g8q01d-App"
    >
      <header
        class="css-1a7bgku-App"
      >
        <h1>
          Fetch Rewards Project
        </h1>
      </header>
      <div
        class="css-1yuj46c-TextInput"
      >
        <label
          class="css-5e23jr-TextInput"
          data-testid="firstnameinput-label"
          for="nameinput"
        >
          First Name: 
        </label>
        <input
          class="css-1a9jkjf-TextInput"
          data-testid="firstnameinput-input"
          id="nameinput"
          name="name"
          type="text"
          value=""
        />
      </div>
      <div
        class="css-1yuj46c-TextInput"
      >
        <label
          class="css-5e23jr-TextInput"
          data-testid="middlenameinput-label"
          for="nameinput"
        >
          Middle Name: 
        </label>
        <input
          class="css-1a9jkjf-TextInput"
          data-testid="middlenameinput-input"
          id="nameinput"
          name="name"
          type="text"
          value=""
        />
      </div>
      <div
        class="css-1yuj46c-TextInput"
      >
        <label
          class="css-5e23jr-TextInput"
          data-testid="lastnameinput-label"
          for="nameinput"
        >
          Last Name: 
        </label>
        <input
          class="css-1a9jkjf-TextInput"
          data-testid="lastnameinput-input"
          id="nameinput"
          name="name"
          type="text"
          value=""
        />
      </div>
      <div
        class="css-1yuj46c-TextInput"
      >
        <label
          class="css-5e23jr-TextInput"
          data-testid="emailinput-label"
          for="emailinput"
        >
           Email: 
        </label>
        <input
          class="css-1a9jkjf-TextInput"
          data-testid="emailinput-input"
          id="emailinput"
          name="email"
          type="text"
          value=""
        />
      </div>
      <div
        class="css-1yuj46c-TextInput"
      >
        <label
          class="css-5e23jr-TextInput"
          data-testid="passwordinput-label"
          for="passwordinput"
        >
           Password: 
        </label>
        <input
          class="css-1a9jkjf-TextInput"
          data-testid="passwordinput-input"
          id="passwordinput"
          name="password"
          type="text"
          value=""
        />
      </div>
      <div
        class="css-1xfx4av-App"
      >
        <button
          class="css-6lzom5-App"
        >
          Submit
        </button>
      </div>
      <div>
        <h1
          class="css-1a7bgku-App"
        >
          Idle...
        </h1>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
